typescript/src/redemption.ts

```diff
@@ -0,0 +1,291 @@
+ * the redemption), an output for each redemption request provided, and a change
+ * output if the redemption requests do not consume the entire amount of the
+ * single input.
+ * @dev The caller is responsible for ensuring the redemption request list is
+ *      correctly formed:
+ *        - there is at least one redemption
+ *        - the `requestedAmount` in each redemption request is greater than
+ *          the sum of its `txFee` and `treasuryFee`
+ *        - the redeemer address in each redemption request is of a standard
+ *          type (P2PKH, P2WPKH, P2SH, P2WSH).
+ * @param walletPrivateKey  - The private key of the wallet in the WIF format
+ * @param mainUtxo - The main UTXO of the wallet. Must match the main UTXO held
+ *        by the on-chain Bridge contract
+ * @param redemptionRequests - The list of redemption requests
+ * @param witness - The parameter used to decide the type of the change output.
+ *                  P2WPKH if `true`, P2PKH if `false`
+ * @returns Bitcoin redemption transaction in the raw format.
+ */
+export async function createRedemptionTransaction(
+  walletPrivateKey: string,
+  mainUtxo: UnspentTransactionOutput & RawTransaction,
+  redemptionRequests: RedemptionRequest[],
+  witness: boolean
+): Promise<RawTransaction> {
+  if (redemptionRequests.length < 1) {
+    throw new Error("There must be at least one request to redeem")
+  }
+
+  const walletKeyRing = createKeyRing(walletPrivateKey, witness)
+  const walletAddress = walletKeyRing.getAddress("string")
+
+  // Use the main UTXO as the single transaction input
+  const inputCoins = [
+    bcoin.Coin.fromTX(
+      bcoin.MTX.fromRaw(mainUtxo.transactionHex, "hex"),
+      mainUtxo.outputIndex,
+      -1
+    ),
+  ]
+
+  const transaction = new bcoin.MTX()
+
+  let txTotalFee = 0
+  let totalOutputsValue = 0
+
+  // Process the requests
+  for (const request of redemptionRequests) {
+    // Calculate the value of the output by subtracting tx fee and treasury
+    // fee for this particular output from the requested amount
+    const outputValue = request.requestedAmount
+      .sub(request.txFee)
+      .sub(request.treasuryFee)
+
+    // Add the output value to the total output value
+    totalOutputsValue += outputValue.toNumber()
+
+    // Add the fee for this particular request to the overall transaction fee
+    txTotalFee += request.txFee.toNumber()
+
+    // Only allow standard address type to receive the redeemed Bitcoins
+    const address = bcoin.Address.fromString(request.redeemerAddress)
+    if (
+      address.isPubkeyhash() ||
+      address.isWitnessPubkeyhash() ||
+      address.isScripthash() ||
+      address.isWitnessScripthash()
+    ) {
+      transaction.addOutput({
+        script: bcoin.Script.fromAddress(address),
+        value: outputValue.toNumber(),
+      })
+    } else {
+      throw new Error("Redemption address must be P2PKH, P2WPKH, P2SH or P2WSH")
+    }
+  }
```

Comment:
Step 7: This validation will be no longer needed. Addresses should be already validated at the redemption request stage. Regarding output script, we can create a `bcoin.Script` instance based on the given request's `redeemerOutputScript` field.

Reply:
Done in https://github.com/keep-network/tbtc-v2/pull/199/commits/4b3be0a5f176ab3e5733f1876ff60fa6d3ca53a2.